src/lib/support/BufBound.h

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    basic double-ended queue, intended to be embedded as a member
+ *     of an object
+ */
+
+#ifndef CHIP_BUFBOUND_H
+#define CHIP_BUFBOUND_H
+
+#include <stdint.h>
+#include <string.h>
+
+namespace chip {
+
+/**
+ *  @class BufBound
+ *
+ *  simple buffer writer
+ */
+class BufBound
+{
+private:
+    uint8_t * mBuf;
+    size_t mLen;
+    size_t mWritten;
+
+public:
+    BufBound(uint8_t * buf, size_t len) : mBuf(buf), mLen(len), mWritten(0) {}
+
+    /*
+     * @brief append a single byte
+     */
+    size_t Put(uint8_t c)
+    {
+        if (mWritten < mLen)
+        {
+            mBuf[mWritten] = c;
+        }
+        return ++mWritten;
+    }
+
+    /*
+     * @brief append a null terminated string, exclude the null term
+     */
+    size_t Put(const char * s)
+    {
+        while (*s != 0)
+        {
+            Put(*s++);
+        }
+        return mWritten;
+    }
+
+    size_t PutLE(uint64_t x, size_t size)
```

Comment:
We should just reuse nlio here rather than duplicating the logic there.

Reply:
nlio doesn't support not writing if the buffer is exhausted

Reply:
Is the long-term intent that this API is used to write machine-sized quantities across buffer boundaries to handle heads and tails?

Reply:
I'm hoping in the long term folks stop copy-pasting memmove(), length checks, etc. for serialization code and think about abstraction.  This kind of abstraction has worked very well for other past projects.

Using BufBound, the same code can be used for safe serialization and measurement.  Nlio has no facility for this, AFAICT.  The measuring part must be written separately, or alongside.